REGRESSION (285925@main): [iOS] Selection highlights for non-editable content sometimes disappear

whsieh opened the pull request:
#### 4aac13b0792324cd39da9b593980e2c79c11e5ce
<pre>
REGRESSION (285925@main): [iOS] Selection highlights for non-editable content sometimes disappear
<a href="https://bugs.webkit.org/show_bug.cgi?id=286490">https://bugs.webkit.org/show_bug.cgi?id=286490</a>
<a href="https://rdar.apple.com/143296175">rdar://143296175</a>

Reviewed by Abrar Rahman Protyasha.

After the changes in <a href="https://commits.webkit.org/285925@main">https://commits.webkit.org/285925@main</a>, we now apply the &quot;hidden selection
container&quot; heuristic to non-editable, non-collapsed selection ranges as well as editable selections
(e.g. Google Docs). This worked by first finding the common ancestor `Node` that contains both
selection endpoints, and then checking if that container&apos;s renderer is either in a completely- or
nearly-transparent layer, or is otherwise completely clipped out (based on visible repaint rects).

However, this logic breaks in the case where the selection endpoints&apos; common ancestor in the render
tree is _not_ the same as the common ancestor in the DOM — for example, when using the inline `font`
element like so:

```
&lt;font face=&quot;…&quot;&gt;
    &lt;p&gt;abc&lt;/p&gt; &lt;!-- selection starts here --&gt;
    …
    &lt;p&gt;def&lt;/p&gt; &lt;!-- selection ends here --&gt;
&lt;/font&gt;
```

In this case, the `font` element is common ancestor in the DOM, but an anonymous block renderer is
the common ancestor in the actual render tree. This causes us to incorrectly treat this selection as
fully clipped, based on the fact that the inline `font` element renderer is empty. To fix this, we
adjust the heuristic to find the common ancestor in the render tree that contains both selection
endpoints, and then check if that common ancestor is hidden.

* LayoutTests/editing/selection/ios/do-not-hide-selection-in-non-editable-visible-container-expected.txt: Added.
* LayoutTests/editing/selection/ios/do-not-hide-selection-in-non-editable-visible-container.html: Added.

Add a new layout test to exercise this fix.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::hasEmptyVisibleRectRespectingParentFrames const):
(WebCore::RenderObject::hasNonEmptyVisibleRectRespectingParentFrames const): Deleted.

Drive-by fix: remove the `Non` from this method, which returns true if and only if the renderer&apos;s
(or one of its parent frames&apos;) visible rect is empty. Note that the only call site expects this
behavior, so the resulting behavior is correct (the method name is just misleading).

* Source/WebCore/rendering/RenderObject.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::isTransparentOrFullyClipped const):
(WebKit::closestCommonContainerInRenderTree):

Add a helper method to return the deepest common renderer that contains both selection endpoints.

(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/289369@main">https://commits.webkit.org/289369@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d491e33129ab6051580679e6b263fd32b3a82d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6194 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41019 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91532 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37419 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88736 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6462 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67027 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24814 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89690 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4916 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/78490 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47349 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4711 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32822 "Found 5 new test failures: editing/undo/redo-reapply-edit-command-crash.html http/tests/security/cookie-module-propagate.html http/tests/security/cookie-module.html imported/w3c/web-platform-tests/cookies/schemeful-same-site/schemeful-websockets.sub.tentative.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe-loading-lazy-multiple-queued-navigations.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36537 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33707 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93427 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13840 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10035 "Found 1 new test failure: swipe/navigate-event-back-swipe-verify-ua-transition.html (failure)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75829 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14041 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75016 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18457 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19329 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17727 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6625 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13863 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13601 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17045 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15385 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -576,12 +576,49 @@
+        ancestorsOfStart.add(container);
+
+    for (CheckedPtr container = endRenderer->container(); container; container = container->container()) {
+        if (ancestorsOfStart.contains(container))
```
Do we need to consider that `contains()` will not respect the insertion sequence of containers into `ancestorsOfStart`?

whsieh replied:
Hm.. I think it should be fine, since `ancestorsOfStart` is unordered anyways. This is just a way to scan for the deepest renderer containing the end that _also_ contains the start, in `O(d)` time (`d` being the depth of the tree)

whsieh committed:
REGRESSION (285925@main): [iOS] Selection highlights for non-editable content sometimes disappear
https://bugs.webkit.org/show_bug.cgi?id=286490
rdar://143296175

Reviewed by Abrar Rahman Protyasha.

After the changes in https://commits.webkit.org/285925@main, we now apply the "hidden selection
container" heuristic to non-editable, non-collapsed selection ranges as well as editable selections
(e.g. Google Docs). This worked by first finding the common ancestor `Node` that contains both
selection endpoints, and then checking if that container's renderer is either in a completely- or
nearly-transparent layer, or is otherwise completely clipped out (based on visible repaint rects).

However, this logic breaks in the case where the selection endpoints' common ancestor in the render
tree is _not_ the same as the common ancestor in the DOM — for example, when using the inline `font`
element like so:

```
<font face="…">
    <p>abc</p> <!-- selection starts here -->
    …
    <p>def</p> <!-- selection ends here -->
</font>
```

In this case, the `font` element is common ancestor in the DOM, but an anonymous block renderer is
the common ancestor in the actual render tree. This causes us to incorrectly treat this selection as
fully clipped, based on the fact that the inline `font` element renderer is empty. To fix this, we
adjust the heuristic to find the common ancestor in the render tree that contains both selection
endpoints, and then check if that common ancestor is hidden.

* LayoutTests/editing/selection/ios/do-not-hide-selection-in-non-editable-visible-container-expected.txt: Added.
* LayoutTests/editing/selection/ios/do-not-hide-selection-in-non-editable-visible-container.html: Added.

Add a new layout test to exercise this fix.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::hasEmptyVisibleRectRespectingParentFrames const):
(WebCore::RenderObject::hasNonEmptyVisibleRectRespectingParentFrames const): Deleted.

Drive-by fix: remove the `Non` from this method, which returns true if and only if the renderer's
(or one of its parent frames') visible rect is empty. Note that the only call site expects this
behavior, so the resulting behavior is correct (the method name is just misleading).

* Source/WebCore/rendering/RenderObject.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::isTransparentOrFullyClipped const):
(WebKit::closestCommonContainerInRenderTree):

Add a helper method to return the deepest common renderer that contains both selection endpoints.

(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/289369@main
